Source/Core/VideoBackends/Metal/MTLMain.mm

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "VideoBackends/Metal/VideoBackend.h"
+
+#include <AppKit/AppKit.h>
+#include <Metal/Metal.h>
+#include <QuartzCore/QuartzCore.h>
+
+#include "Common/Common.h"
+#include "Common/MsgHandler.h"
+
+#include "VideoBackends/Metal/MTLObjectCache.h"
+#include "VideoBackends/Metal/MTLPerfQuery.h"
+#include "VideoBackends/Metal/MTLRenderer.h"
+#include "VideoBackends/Metal/MTLStateTracker.h"
+#include "VideoBackends/Metal/MTLUtil.h"
+#include "VideoBackends/Metal/MTLVertexManager.h"
+
+#include "VideoCommon/FramebufferManager.h"
+#include "VideoCommon/VideoCommon.h"
+#include "VideoCommon/VideoConfig.h"
+
+std::string Metal::VideoBackend::GetName() const
+{
+  return NAME;
+}
+
+std::string Metal::VideoBackend::GetDisplayName() const
+{
+  // i18n: Apple's Metal graphics API (https://developer.apple.com/metal/)
+  return _trans("Metal");
+}
+
+static bool WindowSystemTypeSupportsMetal(WindowSystemType type)
+{
+  switch (type)
+  {
+  case WindowSystemType::MacOS:
+    return true;
+  default:
+    return false;
+  }
+}
+
+bool Metal::VideoBackend::Initialize(const WindowSystemInfo& wsi)
+{
+  @autoreleasepool
+  {
+    if (!WindowSystemTypeSupportsMetal(wsi.type) || !wsi.render_surface)
+    {
+      PanicAlertFmt("Bad WindowSystemInfo for Metal renderer.");
+      return false;
+    }
+
+    auto devs = Util::GetAdapterList();
+    if (devs.empty())
+    {
+      PanicAlertFmt("No Metal GPUs detected.");
+      return false;
+    }
```

Comment:
I think this won't behave quite right, since `Initialize` is only called when starting a game (and when it returns `false`, a separate panic alert is shown). I think it would be better to override `GetWarningMessage` like is done in D3DMain, and always show a panic alert and return false if there are no adapters.

(Though also, no other backend actually does this check as far as I can tell. I'm not sure how they behave if there are no adapters.)

Reply:
It's a bit more important in Metal because Macs with older GPUs will hit this case.

(Well maybe not as important anymore, 10.14 requires a Metal-supporting GPU, and we just dropped support for 10.13)

> and always show a panic alert and return false if there are no adapters.

Are you saying to leave that panic alert there and add a `GetWarningMessage` implementation, or move it somewhere else and add a `GetWarningMessage` implementation?

Reply:
`GetWarningMessage` is called when changing the backend on the UI and shows a prompt to cancel the changes. My comment about "return false" is nonsense; for some reason I was thinking that `GetWarningMessage` showed the alert itself and decided whether to cancel the change, so I was suggesting always canceling it, but that's not how it works. Instead, you should add a `GetWarningMessage` implementation and then leave the `Initialize` implementation as-is.